File: Arrays/RotateArray/rotateArray.py
from typing import List

class Solution:
    def reversePartOfArray(self, arr: List[int], start: int, end : int) -> None:
        if start < 0 or end >= len(arr): return
        arr[start:end+1] = arr[start:end+1][::-1]

    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        n = len(nums)
        k = k%n
        self.reversePartOfArray(nums,0,n-k-1)    
        self.reversePartOfArray(nums,n-k,n-1)    
        self.reversePartOfArray(nums,0,n-1)

File: Arrays/RotateArray/rotateArray.cpp
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    void reversePartOfArray(int startIdx, int endIdx, vector<int> &nums){
        if(startIdx < 0 || endIdx >= nums.size())return;
        // nums.end()-nums.size() = nums.begin()
        reverse(nums.begin()+startIdx, nums.end()-nums.size()+endIdx+1);
        return;
    }
    
    void rotate(vector<int>& nums, int k) {
        int n = nums.size();
        if(k>n) k=k%n;
        reversePartOfArray(0,n-k-1,nums);
        reversePartOfArray(n-k,n-1,nums);
        reversePartOfArray(0,n-1,nums);
        return;
    }
};